### ISSUE TITLE
Add Security for code

### ISSUE THREAD

**Issue by gladson** [neutralinojs/neutralinojs#153]
If possible I could implement some security for the codes.

[Protect Source Code](http://docs.nwjs.io/en/latest/For%20Users/Advanced/Protect%20JavaScript%20Source%20Code/)

Something like the ASAR of Electron, could be the way, but without the possibility of return.

**Comment by wwderw** [comment]
I was under the impression that ASAR format really wasn't that secure for source protection?  Wouldn't something like how NWjs does their source protection better?  Only other alternative would be WASM, but I don't think emscripten is far enough along to protect the majority of code (at least not in my experience with emscripten).

**Comment by shalithasuranga** [comment]
Now we are using the .asar format. Therefore, I will close this issue. Thanks for reporting the issue/idea.

**Comment by TheJaredWilcurt** [comment]
@wwderw

You are correct. ASAR is just a glorified ZIP file, anyone can Google it and have access to the source code/files in seconds. NW.js's approach is to load your JS file into the V8 engine and then store a snapshot of it running in memory as native code. This is equivalent to a native binary build in other languages, like C. So there is no way to retrieve the original source code. You would need to decompile the snapshot binary (high barrier of entry requiring knowledge of the specific platform it was compiled for and it's memory management specifics) and have very deep knowledge of the V8 engine to start understanding the generated machine code commands. Attempting to map them back to JavaScript code would be a monumental effort as well. To my knowledge there has never been a successful attempt to reverse engineer a V8 snapshot.

In short. If you want to protect your source code, your options are:

* Use [NW.js](https://nwjs.io)
* Use [a tool](https://xpda.net) that builds to native code (using something other than JavaScript for your source)
* Store your secret code on a remote server and access it's functionality over the internet (would mean your application functionality would be limited or not work at all when offline)

**Comment by DEVLOPRR** [comment]
If you are so worried about your source code, write the entire app in C or C++, but even that could be decompiled, the NW.js uses v8 snapshots which can be decompiled too.

v8 snapshots weren't even meant for hiding source code, they were meant to significantly boost the startup time of the code execution.

Worrying about source code is useless, maybe use a protective License.

**Comment by wwderw** [comment]
Well that's what I started doing, writing in C, C++ etc.

It's not so much about about total protection, it's about not quite being low hanging fruit.  Given enough time, there really isn't anything that anyone can do about protection, but I doubt anything that I write would be beyond the time of any nominal script that a person would use on it.

Plus, there are times where 3rd party blobs are used that require the minimum to be used (not total lockdown DRM mind you, but something other than just a virtual file tree).
